pmd-java/src/main/resources/category/java/bestpractices.xml

```diff
@@ -421,12 +421,13 @@ if (rst.next()) {    // result is properly examined and used
     <rule name="ConstantsInInterface"
           language="java"
           since="5.5"
-          message="Avoid constants in interfaces. Interfaces define types, constants are implementation details better placed in classes or enums. See Effective Java, item 19."
+          message="Avoid constants in interfaces. Interfaces define types, constants are implementation details better placed in classes or enums. If the constants are best viewed as members of an enumerated type, you should export them with an enum type.
+            For other scenarios, consider using a utility class. See Effective Java, item 17 (1st edition), item 19 (2nd edition), or item 22 (3rd edition)."
```

Comment:
as discussed in the original issue, I'd use the item name rather than the numbers, as they keep changing from edition to edition…

```suggestion
            For other scenarios, consider using a utility class. See Effective Java's 'Use interfaces only to define types'."
```

Reply:
I really don't think it's a good idea to have a 5-sentence message. I prefer having a short message and enhancing the description. The message shouldn't have to explain the rule.

My rule of thumb when writing rules is to have the message just point out *what* the rule detected. *Why* it's a problem and how to fix it should mostly be left to the description of the rule, especially when the fix is not obvious like here.

I think the message should not be more complex than "This interface define constants, which are better defined in enums or classes."

In the [guidelines](https://pmd.github.io/latest/pmd_devdocs_major_rule_guidelines.html#dosdonts-rule-rules) for standard rules we also recommend avoiding messages that start with "Avoid blablabla", and I think it's a good time to correct that.